components/brave_vpn/browser/connection/wireguard/brave_vpn_wireguard_connection_api_base.cc

```diff
@@ -0,0 +1,143 @@
+
+/* Copyright (c) 2023 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+#include "brave/components/brave_vpn/browser/connection/wireguard/brave_vpn_wireguard_connection_api_base.h"
+
+#include "brave/components/brave_vpn/common/brave_vpn_utils.h"
+#include "brave/components/brave_vpn/common/mojom/brave_vpn.mojom.h"
+#include "brave/components/brave_vpn/common/pref_names.h"
+#include "components/prefs/pref_service.h"
+
+namespace brave_vpn {
+
+using ConnectionState = mojom::ConnectionState;
+
+BraveVPNWireguardConnectionAPIBase::BraveVPNWireguardConnectionAPIBase(
+    scoped_refptr<network::SharedURLLoaderFactory> url_loader_factory,
+    PrefService* local_prefs)
+    : BraveVPNOSConnectionAPI(url_loader_factory, local_prefs) {}
+
+BraveVPNWireguardConnectionAPIBase::~BraveVPNWireguardConnectionAPIBase() =
+    default;
+
+void BraveVPNWireguardConnectionAPIBase::SetSelectedRegion(
+    const std::string& name) {
+  GetRegionDataManager().SetSelectedRegion(name);
+}
+
+void BraveVPNWireguardConnectionAPIBase::OnWireguardServiceLaunched(
+    bool success) {
+  UpdateAndNotifyConnectionStateChange(
+      success ? ConnectionState::CONNECTED : ConnectionState::CONNECT_FAILED);
+}
+
+void BraveVPNWireguardConnectionAPIBase::OnWireguardKeypairGenerated(
+    brave_vpn::wireguard::WireguardKeyPair key_pair) {
+  if (!key_pair.has_value()) {
+    VLOG(1) << __func__ << " : failed to get keypair";
+    UpdateAndNotifyConnectionStateChange(ConnectionState::CONNECT_FAILED);
+    return;
+  }
+  const auto [public_key, private_key] = key_pair.value();
+  GetAPIRequest()->GetWireguardProfileCredentials(
+      base::BindOnce(
+          &BraveVPNWireguardConnectionAPIBase::OnGetProfileCredentials,
+          base::Unretained(this), private_key),
+      GetSubscriberCredential(local_prefs()), public_key, GetHostname());
+}
+
+void BraveVPNWireguardConnectionAPIBase::Connect() {
+  if (GetConnectionState() == ConnectionState::CONNECTED) {
+    Disconnect();
```

Comment:
Curious why call `Disconnect()` here? I think just return is sufficient?

Reply:
We have same for RAS https://github.com/brave/brave-core/blob/master/components/brave_vpn/browser/connection/ikev2/brave_vpn_ras_connection_api_base.cc#L102, seems like to re-set new regions?

Reply:
Then we need more logic such as `need_connect_` and we have to call `Connect()` from `OnDisconnected()` based on `need_connect_`.
I think this PR doesn't have that logic yet so I think just `return` now appropriate?

Reply:
removed Disconnect call, we do not need this because it will be stopped and re-created in the service synchronously